### PR TITLE
Posenet Improvements

### PR DESCRIPTION
**Pull request by oveddan**
*Removed index.ts files and just renamed to decode single pose
* Made demo enable you to change gui and see results of the algorithm parameters in real time.
* fixed build config so that sourcemaps work. renamed to tensorflow/models
* fixed output files in package.json
* fixed how demos would fail to require posenet model by moving .babelrc to the right location
* switched to pull images from cdn
* Moved "developing the demos" instructions to the bottom of readme.
* Changed everywhere in readme to import from `@tensorflow-models/posenet`
* Added return types
* Added instructions for pixel values.
* Changed joint names to be camelCase
* Upgraded to 0.7.2
* casting image data before dividing to fix error
* Fixing a warning around URL. createObjectURL
* Modifying getAdjacentKeyPoints so it returns the name of the keypoints
* Converted argmax2d to use tfjs' integer div after bug was fixed.
* Single Pose estimation should run faster now
* Change keypoints to always return the part name, as a property `part`
* Removing outputVideoResolution in favor of a constant canvas size
* Adding controls to go from single-pose to multi-pose detector
* Changed the color and thickness of skeleton line just to increase contrast a bit more
* Adding stats.js to camera demo to show FPS
* Adding a footer to explain the inputs and controls
* Adding a footer to coco demo and clearing up the UI on first load so it's clear the model is loading
* Small tweak to detect pose button in coco images
* Adding a README to the demos folder as well as a few GIFS
* Enabled posenet to import different checkpoints with different mobilenet architectures.
* Broke out mobilenet functionality into its own class, which is now contained by the PoseNet model.
* Split up converting from input layers to strided and then running convolutions into two distinct steps so that it's easier to read, follow, test and debug.

---
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-models/3)

### CODE REVIEW

**Comment by nsthorat** [comment]
Two high-level API comments. I think these are important because they will set a precedent for other models.

- Can we make the posenet API have no tensors? Take an image element (or anything that fromPixels allows) and returns a javascript array?
- I think we should aim for this API:
```
import posenet from '@tensorflow-models/posenet';
const net = posenet();
await net.load();
await poseNet.estimateSinglePose(resized, outputStride);
```

To do this, you'll have to use a default export (basically a factory method just so that you hide using "new"):
https://developer.mozilla.org/en-US/docs/web/javascript/reference/statements/export#Using_the_default_export

Overall this is awesome. If you want, you can just respond to my nit comments inline and tackle these larger things in another PR (actually, I would prefer that so this PR is smaller).

Nice work again Dan!

---

Reviewed 25 of 68 files at r1.
Review status: 24 of 62 files reviewed at latest revision, all discussions resolved.

---

*[posenet/package.json, line 13 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/3#-LBmfmUdAH72nNIzu26w:-LBmfmUdAH72nNIzu26x:b-734qx1) ([raw file](https://github.com/tensorflow/tfjs-models/blob/da9ff579213505a5f80fa60f666e3ee12dbb18fc/posenet/package.json#L13)):*
> ```JSON
>   },
>   "peerDependencies": {
>     "@tensorflow/tfjs-core": "0.8.1"
> ```

can you make this depend on the union package (tfjs instead of tfjs-core)? I know it's going to bloat the bundle for right now, but unfortunately if we don't do that people won't be able to use tensors created from tfjs union package with posenet.

---

*[posenet/README.md, line 16 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/3#-LBmg8LX3l8hYpEHp94m:-LBmg8LX3l8hYpEHp94n:b-1o4dos) ([raw file](https://github.com/tensorflow/tfjs-models/blob/da9ff579213505a5f80fa60f666e3ee12dbb18fc/posenet/README.md#L16)):*
> ```Markdown
> 
> ```html
> <script src="https://unpkg.com/tfjs"></script>
> ```

this is correct, but just FYI this is not what you're showing in other places (tfjs is the union, tfjs-core is just core).

---

*[posenet/demos/coco.js, line 160 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/3#-LBmn0RD7a8zQbbKl2G_:-LBmn0RD7a8zQbbKl2Ga:bnv2htg) ([raw file](https://github.com/tensorflow/tfjs-models/blob/da9ff579213505a5f80fa60f666e3ee12dbb18fc/posenet/demos/coco.js#L160)):*
> ```JavaScript
> 
> async function decodeMultiplePosesAndDrawResults() {
>   if (!modelOutputs) return;
> ```

use `{}` here

---

*[posenet/demos/coco.js, line 213 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/3#-LBmn2tt5apVS5oB8YzG:-LBmn2tt5apVS5oB8YzH:b-pi9d9n) ([raw file](https://github.com/tensorflow/tfjs-models/blob/da9ff579213505a5f80fa60f666e3ee12dbb18fc/posenet/demos/coco.js#L213)):*
> ```JavaScript
> let guiState;
> 
> // var detectButton = { detect: () => {
> ```

remove this

---

*[posenet/demos/README.md, line 23 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/3#-LBmoMVF9PfbHvXYH02b:-LBmoMVF9PfbHvXYH02c:bn68a4t) ([raw file](https://github.com/tensorflow/tfjs-models/blob/da9ff579213505a5f80fa60f666e3ee12dbb18fc/posenet/demos/README.md#L23)):*
> ```Markdown
> From the `posenet` folder run:
> 
>     yarn
> ```

use triple backticks here for markdown, here and below

---

*[posenet/demos/README.md, line 25 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/3#-LBmoJcB2S5A8uRyerLh:-LBmoJcB2S5A8uRyerLi:b-2vhcqx) ([raw file](https://github.com/tensorflow/tfjs-models/blob/da9ff579213505a5f80fa60f666e3ee12dbb18fc/posenet/demos/README.md#L25)):*
> ```Markdown
>     yarn
> 
> Cd into the demos folder
> ```

cd lowercase

---

*[posenet/src/checkpoints.ts, line 1 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/3#-LBmmAcH6jkd2G9mIlBj:-LBmmAcH6jkd2G9mIlBk:bmckcm4) ([raw file](https://github.com/tensorflow/tfjs-models/blob/da9ff579213505a5f80fa60f666e3ee12dbb18fc/posenet/src/checkpoints.ts#L1)):*
> ```XML
> 
> ```

apache license here

---

*[posenet/src/checkpoints.ts, line 3 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/3#-LBmqep9CyL2-hX-0OWP:-LBmqep9CyL2-hX-0OWQ:bup3jh4) ([raw file](https://github.com/tensorflow/tfjs-models/blob/da9ff579213505a5f80fa60f666e3ee12dbb18fc/posenet/src/checkpoints.ts#L3)):*
> ```XML
> 
> const GOOGLE_CLOUD_STORAGE_DIR =
>     'https://storage.googleapis.com/cl-move-mirror.appspot.com/';
> ```

An example of one of your checkpoints:
https://storage.googleapis.com/tfjs-models/weights/posenet/mobilenet_v1_101

So this should become:
https://storage.googleapis.com/tfjs-models/weights/posenet/

---

*[posenet/src/checkpoints.ts, line 5 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/3#-LBmoTIBDyZIjnr0bh_0:-LBmoTIBDyZIjnr0bh_1:bhj4z4m) ([raw file](https://github.com/tensorflow/tfjs-models/blob/da9ff579213505a5f80fa60f666e3ee12dbb18fc/posenet/src/checkpoints.ts#L5)):*
> ```XML
>     'https://storage.googleapis.com/cl-move-mirror.appspot.com/';
> 
> import {mobileNetArchitectures, ConvolutionDefinition} from './mobilenet'
> ```

put the import above the GCS path

---

*[posenet/src/index.ts, line 21 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/3#-LBn4E6b325w0Rj4tC1Q:-LBn4E6b325w0Rj4tC1R:b-92nhxc) ([raw file](https://github.com/tensorflow/tfjs-models/blob/da9ff579213505a5f80fa60f666e3ee12dbb18fc/posenet/src/index.ts#L21)):*
> ```XML
> import decodeSinglePose from './singlePose/decodeSinglePose';
> 
> export { checkpoints } from './checkpoints';
> ```

make sure you have clang-format installed properly (vscode extension, .vscode in the root of the project). looks like this isn't google style

---

*[posenet/src/mobilenet.ts, line 1 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/3#-LBn6h1UEzhx_rr7G0z0:-LBn6h1UEzhx_rr7G0z1:b-wqhi78) ([raw file](https://github.com/tensorflow/tfjs-models/blob/da9ff579213505a5f80fa60f666e3ee12dbb18fc/posenet/src/mobilenet.ts#L1)):*
> ```XML
> import * as tf from '@tensorflow/tfjs-core';
> ```

license

---

*[posenet/src/mobilenet.ts, line 12 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/3#-LBn6jhp8KVVNynDsXKk:-LBn6jhp8KVVNynDsXKl:b-u0ppu2) ([raw file](https://github.com/tensorflow/tfjs-models/blob/da9ff579213505a5f80fa60f666e3ee12dbb18fc/posenet/src/mobilenet.ts#L12)):*
> ```XML
> const mobileNet100Architecture: ConvolutionDefinition[] = [
>   ['conv2d', 2],
>   ['seperableConv', 1],
> ```

s/seperable/separable

here and below

---

*[posenet/src/mobilenet.ts, line 77 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/3#-LBn6qdAAprJw3LsEYv3:-LBn6qdAAprJw3LsEYv4:b3lg3v8) ([raw file](https://github.com/tensorflow/tfjs-models/blob/da9ff579213505a5f80fa60f666e3ee12dbb18fc/posenet/src/mobilenet.ts#L77)):*
> ```XML
> };
> 
> /* Takes a mobilenet architectures' convolution definitions and converts them
> ```

use standard jsdoc with `/**` on its own line

---

*[posenet/src/posenet_test.ts, line 30 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/3#-LBn77xa6c_wtgpWh1Q5:-LBn77xa6c_wtgpWh1Q6:b-mb9903) ([raw file](https://github.com/tensorflow/tfjs-models/blob/da9ff579213505a5f80fa60f666e3ee12dbb18fc/posenet/src/posenet_test.ts#L30)):*
> ```XML
> 
>   describe('estimateSinglePose', () => {
>     it('does not leak memory', async function() {
> ```

okay, so unfortunately jasmine is really wonky with async unit tests. If this fails for some reason you will only get "async callback not called". Check out what we do here:
https://github.com/tensorflow/tfjs-core/blob/master/src/io/weights_loader_test.ts#L81

TLDR: you can't use async functions, you have to use .then :/

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-models/3#-:-LBmmHbM1k-SzWXldfMJ:bm6lr4s)*

**Comment by nsthorat** [comment]
Thinking about these a little more, actually I think we should even just making the weights load when posenet is created:

```
import posenet from '@tensorflow-models/posenet';
const net = await posenet('1.0'); // this calls load internally to load all the weights
const result = await poseNet.estimateSinglePose(imgElement);  // outputStride is optional
```

The idea being to reduce lines of code and avoid people forgetting to call load.

---

Review status: 23 of 62 files reviewed at latest revision, 14 unresolved discussions.

---

*[posenet/src/posenet.ts, line 32 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/3#-LBn9XPs-h7HDmJ2a9C0:-LBn9XPs-h7HDmJ2a9C1:b1beuqp) ([raw file](https://github.com/tensorflow/tfjs-models/blob/da9ff579213505a5f80fa60f666e3ee12dbb18fc/posenet/src/posenet.ts#L32)):*
> ```XML
> 
>   async load(
>       checkpointUrl: string = defaultCheckpoint.url,
> ```

are you expecting a user to pass something here? this seems a little funny.

maybe the different configurations can be passed to the constructor function as a union type instead of as a URL (which the user wouldn't know)? And when you do that, can you make sure you do runtime checks for valid arguments with a good error message?

---

*[posenet/src/posenet.ts, line 79 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/3#-LBnBV4f-OjL4zQTSs-e:-LBnBV4f-OjL4zQTSs-f:bcrysmu) ([raw file](https://github.com/tensorflow/tfjs-models/blob/da9ff579213505a5f80fa60f666e3ee12dbb18fc/posenet/src/posenet.ts#L79)):*
> ```XML
>    * @return heatmapScores, offsets, displacementFwd, displacementBwd
>    */
>   predictForMultiPose(input: tf.Tensor3D, outputStride: OutputStride = 32): {
> ```

can you do a runtime check for the value of outputStride being valid? otherwise JS users will call this and get a cryptic error. Same for the other method.

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-models/3#-:-LBnA9oZ5j9LMbGDEd6N:b-urqscy)*

**Comment by nsthorat** [comment]
Correction on line 2: instead of '1.0' maybe just let it be a floating point number.
Correction on the last line: s/poseNet/net

---

Review status: 23 of 62 files reviewed at latest revision, 16 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-models/3#-:-LBnD3Nc3ZM5ayi3ORgR:b-y741b9)*

**Comment by nsthorat** [comment]
Review status: 23 of 62 files reviewed at latest revision, 16 unresolved discussions.

---

*[posenet/package.json, line 44 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/3#-LBnHG4s5VdZPQq67xjw:-LBnHG4s5VdZPQq67xjx:b-sgsoo8) ([raw file](https://github.com/tensorflow/tfjs-models/blob/da9ff579213505a5f80fa60f666e3ee12dbb18fc/posenet/package.json#L44)):*
> ```JSON
>   },
>   "browserify-shim": {
>     "tf": "global:dl"
> ```

s/dl/tf

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-models/3)*

**Comment by oveddan** [comment]
Totally agree with doing something like `await posenet`

What does the 1.0 mean?

And yeah I'll do this change in a separate PR

---

Review status: 23 of 62 files reviewed at latest revision, 17 unresolved discussions.

---

*[posenet/package.json, line 13 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/3#-LBmfmUdAH72nNIzu26w:-LBpso_C7tNvfZlakAS8:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-models/blob/da9ff579213505a5f80fa60f666e3ee12dbb18fc/posenet/package.json#L13)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

can you make this depend on the union package (tfjs instead of tfjs-core)? I know it's going to bloat the bundle for right now, but unfortunately if we don't do that people won't be able to use tensors created from tfjs union package with posenet.
</blockquote></details>

Done.

---

*[posenet/package.json, line 44 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/3#-LBnHG4s5VdZPQq67xjw:-LBpso-S8lMjvxIlfOJt:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-models/blob/da9ff579213505a5f80fa60f666e3ee12dbb18fc/posenet/package.json#L44)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

s/dl/tf
</blockquote></details>

Done.

---

*[posenet/demos/coco.js, line 160 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/3#-LBmn0RD7a8zQbbKl2G_:-LBq4UQGAp5tNE5qc9cQ:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-models/blob/da9ff579213505a5f80fa60f666e3ee12dbb18fc/posenet/demos/coco.js#L160)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

use `{}` here
</blockquote></details>

Done.

---

*[posenet/demos/coco.js, line 213 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/3#-LBmn2tt5apVS5oB8YzG:-LBq4zeI-Pw1Ua15uwGe:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-models/blob/da9ff579213505a5f80fa60f666e3ee12dbb18fc/posenet/demos/coco.js#L213)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

remove this
</blockquote></details>

Done.

---

*[posenet/demos/README.md, line 23 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/3#-LBmoMVF9PfbHvXYH02b:-LBq5u9v0FtsPeFthIhv:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-models/blob/da9ff579213505a5f80fa60f666e3ee12dbb18fc/posenet/demos/README.md#L23)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

use triple backticks here for markdown, here and below
</blockquote></details>

Done.

---

*[posenet/demos/README.md, line 25 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/3#-LBmoJcB2S5A8uRyerLh:-LBq61aA7_sV8Um1TmA7:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-models/blob/da9ff579213505a5f80fa60f666e3ee12dbb18fc/posenet/demos/README.md#L25)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

cd lowercase
</blockquote></details>

Done.

---

*[posenet/src/checkpoints.ts, line 1 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/3#-LBmmAcH6jkd2G9mIlBj:-LBpu_91BR_QqwikuMWo:bs7wuop) ([raw file](https://github.com/tensorflow/tfjs-models/blob/da9ff579213505a5f80fa60f666e3ee12dbb18fc/posenet/src/checkpoints.ts#L1)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

apache license here
</blockquote></details>

Done.`

---

*[posenet/src/checkpoints.ts, line 3 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/3#-LBmqep9CyL2-hX-0OWP:-LBpyQKa9aTkNA7ynPuM:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-models/blob/da9ff579213505a5f80fa60f666e3ee12dbb18fc/posenet/src/checkpoints.ts#L3)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

An example of one of your checkpoints:
https://storage.googleapis.com/tfjs-models/weights/posenet/mobilenet_v1_101

So this should become:
https://storage.googleapis.com/tfjs-models/weights/posenet/
</blockquote></details>

Done.

---

*[posenet/src/checkpoints.ts, line 5 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/3#-LBmoTIBDyZIjnr0bh_0:-LBpud4hCDDps_VoaQtF:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-models/blob/da9ff579213505a5f80fa60f666e3ee12dbb18fc/posenet/src/checkpoints.ts#L5)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

put the import above the GCS path
</blockquote></details>

Done.

---

*[posenet/src/index.ts, line 21 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/3#-LBn4E6b325w0Rj4tC1Q:-LBpv12S6GLdN85Px51M:bb7f2mf) ([raw file](https://github.com/tensorflow/tfjs-models/blob/da9ff579213505a5f80fa60f666e3ee12dbb18fc/posenet/src/index.ts#L21)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

make sure you have clang-format installed properly (vscode extension, .vscode in the root of the project). looks like this isn't google style
</blockquote></details>

done

---

*[posenet/src/mobilenet.ts, line 1 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/3#-LBn6h1UEzhx_rr7G0z0:-LBpvAMo7-txsxTsmkXc:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-models/blob/da9ff579213505a5f80fa60f666e3ee12dbb18fc/posenet/src/mobilenet.ts#L1)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

license
</blockquote></details>

Done.

---

*[posenet/src/mobilenet.ts, line 12 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/3#-LBn6jhp8KVVNynDsXKk:-LBpvMgLE6oeyRYVddv1:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-models/blob/da9ff579213505a5f80fa60f666e3ee12dbb18fc/posenet/src/mobilenet.ts#L12)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

s/seperable/separable

here and below
</blockquote></details>

Done.

---

*[posenet/src/mobilenet.ts, line 77 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/3#-LBn6qdAAprJw3LsEYv3:-LBpvUm27NxcsWRHMYDm:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-models/blob/da9ff579213505a5f80fa60f666e3ee12dbb18fc/posenet/src/mobilenet.ts#L77)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

use standard jsdoc with `/**` on its own line
</blockquote></details>

Done.

---

*[posenet/src/posenet.ts, line 32 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/3#-LBn9XPs-h7HDmJ2a9C0:-LBpvcLS7vEK9r9WRqal:b-plgowb) ([raw file](https://github.com/tensorflow/tfjs-models/blob/da9ff579213505a5f80fa60f666e3ee12dbb18fc/posenet/src/posenet.ts#L32)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

are you expecting a user to pass something here? this seems a little funny.

maybe the different configurations can be passed to the constructor function as a union type instead of as a URL (which the user wouldn't know)? And when you do that, can you make sure you do runtime checks for valid arguments with a good error message?
</blockquote></details>

Yeah I wasn't sure what to do here.  You think it should be an architecture name (such as `mobilenet_v1_101` or `mobilenet_v1_050` or id (such as `101` `50`) ? Or an object with a url and a checkpoint architecture?  The advantage of the latter is the user could switch the url.

Let me know what you think is best.

---

*[posenet/src/posenet.ts, line 79 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/3#-LBnBV4f-OjL4zQTSs-e:-LBpw4gwADULV5Anas2Q:b-ycpq2a) ([raw file](https://github.com/tensorflow/tfjs-models/blob/da9ff579213505a5f80fa60f666e3ee12dbb18fc/posenet/src/posenet.ts#L79)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

can you do a runtime check for the value of outputStride being valid? otherwise JS users will call this and get a cryptic error. Same for the other method.
</blockquote></details>

Totally agree with this.

Done.

---

*[posenet/src/posenet_test.ts, line 30 at r1](https://reviewable.io/reviews/tensorflow/tfjs-models/3#-LBn77xa6c_wtgpWh1Q5:-LBq1c2L3iC_DUwgPl0X:bm5qo2j) ([raw file](https://github.com/tensorflow/tfjs-models/blob/da9ff579213505a5f80fa60f666e3ee12dbb18fc/posenet/src/posenet_test.ts#L30)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

okay, so unfortunately jasmine is really wonky with async unit tests. If this fails for some reason you will only get "async callback not called". Check out what we do here:
https://github.com/tensorflow/tfjs-core/blob/master/src/io/weights_loader_test.ts#L81

TLDR: you can't use async functions, you have to use .then :/
</blockquote></details>

Ahhh thats too bad.  I switched to the done syntax.

Maybe we should switch to mocha in the future.

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-models/3#-:-LBprm2W9OHhWWvHdnrj:b-kobh1v)*

**Comment by oveddan** [comment]
Hey so I went through most of this refactoring.  It's mostly good.  Couple issues:

It would be challenging to have that function accept a floating point value, when the checkpoints would be indexed by the multiplier.

Currently I have it like:

```typescript
export const checkpoints: {[multiplier: string]: Checkpoint} = {
  '1.01': {
    url: GOOGLE_CLOUD_STORAGE_DIR + 'mobilenet_v1_101/',
    architecture: mobileNetArchitectures[100]
  },
  '1.00': {
    url: GOOGLE_CLOUD_STORAGE_DIR + 'mobilenet_v1_100/',
    architecture: mobileNetArchitectures[100]
  },
  '0.75': {
    url: GOOGLE_CLOUD_STORAGE_DIR + 'mobilenet_v1_075/',
    architecture: mobileNetArchitectures[75]
  },
  '0.50': {
    url: GOOGLE_CLOUD_STORAGE_DIR + 'mobilenet_v1_050/',
    architecture: mobileNetArchitectures[50]
  }
}
```

It wouldn't be possible to do that with a floating point number, unless we did some conversion and padded 1.0 with a 0.

Otherwise we would have to have a bunch of `if/else` or `case` to look up the proper checkpoint.

Exporting a default from the `posenet` library makes it challenging to import non-default things via `import * as posenet` because you can't do:

```typescript
import posenet, * as posenet from '@tensorflow-models/posenet'
```

What do you think of doing:

```typescript
const net = await posenet.loadNet('1.01');
```

I feel that would be clearer to understand and easier for documentation.

---

Review status: 23 of 62 files reviewed at latest revision, 17 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-models/3#-:-LBqiauf4eOg62t59KHR:b-7vw5te)*
